consensus/src/quorum_store/proof_manager.rs

```diff
@@ -1,59 +1,175 @@
 use futures::StreamExt;
 use futures_channel::mpsc::Receiver;
-use std::collections::HashSet;
+use rand::{seq::SliceRandom, thread_rng};
+use std::{
+    cmp::min,
+    collections::{BTreeMap, HashMap, HashSet},
+    sync::Arc,
+};
 
 #[derive(Debug)]
 pub enum ProofManagerCommand {
     ReceiveProofs(ProofOfStoreMsg),
+    ReceiveBatches(Vec<BatchInfo>),
     CommitNotification(u64, Vec<BatchInfo>),
     Shutdown(tokio::sync::oneshot::Sender<()>),
 }
 
+pub struct BatchQueue {
+    batch_store: Arc<BatchStore>,
+    // Queue per peer to ensure fairness between peers and priority within peer
+    author_to_batches: HashMap<PeerId, BTreeMap<BatchSortKey, BatchInfo>>,
+}
+
+impl BatchQueue {
+    pub fn new(batch_store: Arc<BatchStore>) -> Self {
+        Self {
+            batch_store,
+            author_to_batches: HashMap::new(),
+        }
+    }
+
+    pub fn add_batches(&mut self, batches: Vec<BatchInfo>) {
+        for batch in batches.into_iter() {
+            let queue = self.author_to_batches.entry(batch.author()).or_default();
+            queue.insert(BatchSortKey::from_info(&batch), batch.clone());
+        }
+    }
+
+    pub fn remove_batch(&mut self, batch: &BatchInfo) {
+        if let Some(batch_tree) = self.author_to_batches.get_mut(&batch.author()) {
+            batch_tree.remove(&BatchSortKey::from_info(batch));
+        }
+    }
+
+    pub fn remove_expired_batches(&mut self) {
+        let authors = self.author_to_batches.keys().cloned().collect::<Vec<_>>();
+        for author in authors {
+            if let Some(batch_tree) = self.author_to_batches.get_mut(&author) {
+                batch_tree.retain(|_batch_key, batch| !batch.is_expired());
+            }
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.author_to_batches
+            .values()
+            .map(|batch_tree| batch_tree.len())
+            .sum()
+    }
+
+    pub fn pull_batches(
+        &mut self,
+        max_txns: u64,
+        max_bytes: u64,
+        excluded_batches: Vec<BatchInfo>,
+    ) -> Vec<(BatchInfo, Vec<SignedTransaction>)> {
+        let mut result: Vec<(BatchInfo, Vec<SignedTransaction>)> = vec![];
+        let mut num_txns = 0;
+        let mut num_bytes = 0;
+        let mut iters = vec![];
+        let mut full = false;
+        for (_, batches) in self.author_to_batches.iter() {
+            iters.push(batches.iter().rev());
+        }
+        while !iters.is_empty() {
+            iters.shuffle(&mut thread_rng());
+            iters.retain_mut(|iter| {
+                if full {
+                    return false;
+                }
+                if let Some((_sort_key, batch)) = iter.next() {
+                    if excluded_batches.contains(batch) {
+                        true
+                    } else if num_txns + batch.num_txns() <= max_txns
+                        && num_bytes + batch.num_bytes() <= max_bytes
+                    {
+                        if let Ok(mut persisted_value) =
```

Comment:
Can we log warn! if it's not found? Shouldn't really happen often, so if does maybe having logs will help.

Reply:
Done.